lib/network.nix

```diff
@@ -0,0 +1,317 @@
+{lib}: let
+  /*
+  Converts an IP address from a list of ints to a string.
+
+  Type: prettyIp :: [ Int ] -> String
+
+  Examples:
+    prettyIp [ 192 168 70 9 ]
+    => "192.168.70.9"
+  */
+  prettyIp = addr:
+    lib.concatStringsSep "." (builtins.map builtins.toString addr);
+
+  /*
+  Given a bit mask, return the associated subnet mask.
+
+  Type: bitMaskToSubnetMask :: Int -> [ Int ]
+
+  Examples:
+    bitMaskToSubnetMask 15
+    => [ 255 254 0 0 ]
+    bitMaskToSubnetMask 24
+    => [ 255 255 255 0 ]
+  */
+  bitMaskToSubnetMask = bitMask: let
+    numOctets = 4;
+    octetBits = 8;
+    octetMin = 0;
+    octetMax = 255;
+    # How many initial parts of the mask are full (=255)
+    fullParts = bitMask / octetBits;
+  in
+    lib.genList (
+      idx:
+      # Fill up initial full parts
+        if idx < fullParts
+        then octetMax
+        # If we're above the first non-full part, fill with 0
+        else if fullParts < idx
+        then octetMin
+        # First non-full part generation
+        else _genPartialMask (lib.mod bitMask octetBits)
+    )
+    numOctets;
+
+  /*
+  Generate a the partial portion of a subnet mask.
+
+  Type: _genPartialMask :: Int -> Int
+
+  Examples:
+    _genPartialMask 0
+    => 0
+    _genPartialMask 1
+    => 128
+    _genPartialMask 2
+    => 192
+    _genPartialMask 3
+    => 224
+    _genPartialMask 4
+    => 240
+    _genPartialMask 5
+    => 248
+    _genPartialMask 6
+    => 252
+    _genPartialMask 7
+    => 254
+  */
+  _genPartialMask = n:
+    if n == 0
+    then 0
+    else _genPartialMask (n - 1) / 2 + 128;
+
+  /*
+  Given a subnet mask, return the associated bit mask.
+
+  Type: subnetMaskToBitMask :: [ Int ] -> Int
+
+  Examples:
+    subnetMaskToBitMask [ 255 254 0 0 ]
+    => 15
+    subnetMaskToBitMask [ 255 255 255 0 ]
+    => 24
+  */
+  subnetMaskToBitMask = subnetMask: let
+    partialBits = octet:
+      if octet == 0
+      then 0
+      else (lib.mod octet 2) + partialBits (octet / 2);
+  in
+    builtins.foldl'
+    (x: y: x + y)
+    0
+    (builtins.map partialBits subnetMask);
+
+  /*
+  Given a CIDR, return the IP Address.
+
+  Type: cidrToIpAddress :: String -> [ Int ]
+
+  Examples:
+    cidrToIpAddress "192.168.70.9/15"
+    => [ 192 168 70 9 ]
+  */
+  cidrToIpAddress = cidr: let
+    splitParts = lib.splitString "/" cidr;
+    addr = lib.elemAt splitParts 0;
+    parsed =
+      builtins.map
+      lib.toInt
+      (builtins.match "([0-9]+)\\.([0-9]+)\\.([0-9]+)\\.([0-9]+)" addr);
+    checkBounds = octet:
+      (octet >= 0) && (octet <= 255);
+  in
+    if (builtins.all checkBounds parsed)
+    then parsed
+    else builtins.throw "IP ${prettyIp addr} has out of bounds octet(s)";
+
+  /*
+  Given a CIDR, return the bitmask.
+
+  Type: cidrToBitMask :: String -> Int
+
+  Examples:
+    cidrToBitMask "192.168.70.9/15"
+    => 15
+  */
+  cidrToBitMask = cidr: let
```

Comment:
How certain are you about the term *bitmask*? I would expect `"11111111111111110000000000000000"`.

In IPv6 lingo that number after the slash is named *prefix length*.